### ISSUE TITLE
Absolute paths in 'files' section of JSON query

### ISSUE THREAD

**Issue by puremourning** [ternjs/tern#690]
Hi there,

TL;DR - Adding tern support to YouCompleteMe, need to use absolute paths for the 'files' part of the query. Should this work? Can it?

First a bit of background: I'm a maintainer of [YouCompleteMe](http://github.com/Valloric/YouCompleteMe), and I'm currently working on adding javascript semantic support to [ycmd](http://gitbub.com/Valloric/ycmd), the multi-language code comprehension server that powers it. Tern is very cool and we're really excited about integrating it thanks to your super simple and well-documented, REST API. I got the basics including code completion, docs, goto, etc. working in just a couple of hours.

However, in testing I've noticed a snag related to dirty buffers. When issuing requests, I'm specifying absolute paths in the 'files' object, as ycmd has no knowledge about the project structure. I expected that the server would be able to resolve the absolute path to the relative paths that it uses internally (if that is indeed the case).

This appears to be resolved by [tern_for_vim](http://github.com/ternjs/tern_for_vim) by making all paths relative to the (assumed) project directory, but as a user of the API, I don't really want to have to know anything about tern's project structure.

To be a specific, consider the following filesystem structure, rooted under `/Users/ben/Development/tmp/js`:

```
/Users/ben/Development/tmp/js/lib/a_module.js
/Users/ben/Development/tmp/js/test_file.js
/Users/ben/Development/tmp/js/.tern-project
```

With these contents:

- `./lib/a_module.js`:

```javascript
define( [], function() {
    return {
        /**
         * This function takes a number and does some math on it
         */
        do_math: function( how_much ) {
            return how_much * 0.001;
        }
    }
} )
```

- `./test_file.js`

```javascript
define( [ 'lib/a_module.js' ], function( a_module ) {
  a_module.do_math( 10 );
} );
```

- `./.tern-project`

```json
{
  "libs": [
  ],
  "plugins": {
    "requirejs": {
      "baseURL": "./",
      "paths": {}
    }
  }
}
```

So the idea is pretty simple - a requirejs module which exports an object with a single function in it.

If I make a change to the `lib/a_module.js` and don't save it, ycmd can send a request containing the dirty buffer. This is great, but I'm seeing a problem when I do this, the problem is that the updated (dirty) code is not used. For example if I change it like this:

```javascript
define( [], function() {
    return {
        /**
         * This function takes a number and does some math on it.
         * And some more text.
         */
        do_math: function( how_much ) {
            return how_much * 0.001;
        }
    }
} )
```

And issue the following request:

```json
Request: {
  "files": [
    {
      "text": "define( [ 'lib/a_module.js' ], function( a_module ) {\n    a_module.do_math( 10 );\n} );\n\n",
      "type": "full",
      "name": "/Users/ben/Development/tmp/js/test_file.js"
    },
    {
      "text": "define( [], function() {\n    return {\n        /**\n         * This function takes a number and does some math on it\n         *\n         * And some more text\n         */\n        do_math: function( how_much ) {\n            return how_much * 0.001;\n        }\n    }\n} )\n\n",
      "type": "full",
      "name": "/Users/ben/Development/tmp/js/lib/a_module.js"
    }
  ],
  "query": {
    "end": {
      "line": 1,
      "ch": 13
    },
    "lineCharPositions": true,
    "type": "documentation",
    "file": "/Users/ben/Development/tmp/js/test_file.js",
    "docFormat": "full"
  },
  "timeout": 500
}
```

I get the following reply:

```json
Response: {
  "doc": "This function takes a number and does some math on it",
  "type": "fn(how_much: number) -> number",
  "origin": "lib/a_module.js"
}
```

But I expect something like

```json
Response: {
  "doc": "This function takes a number and does some math on it\n\nAnd some more text",
  "type": "fn(how_much: number) -> number",
  "origin": "lib/a_module.js"
}
```

To check, if I issue the 'files' query, I get this:

```json
Response: {
  "files": [
    "/Users/ben/Development/tmp/js/lib/a_module.js",
    "/Users/ben/Development/tmp/js/test_file.js",
    "lib/a_module.js"
  ]
}
```

I looked a bit at the `tern.js` code and it _seems_ to try and resolve this, but it isn't working for me:

```javascript
    normalizeFilename: function(name) {
      var norm = this.options.normalizeFilename(name).replace(/\\/g, "/")
      if (norm.indexOf(this.projectDir) == 0) norm = norm.slice(this.projectDir.length)
      return norm
    }
```

So I guess the questions are:

- are absolute paths supported for the 'files' part of the query document?
- if so, any advice as to what I am doing wrong?
- if not, what are the chances of supporting this?

As always, let me know if I can provide any more data, examples, etc. 

I am using github checkout, master at d91740fa9cfd2194bcf54c4c9d054b31fe5a5a66

Many thanks,
Ben

**Comment by puremourning** [comment]
FYI, FWIW changing the following seems to sort of make it work for me, but I have no clue if this is safe, sane in any other context:

From:

```javascript
  function ensureFile(srv, name, parent, text) {
    var known = srv.findFile(name);
```

to

```javascript
  function ensureFile(srv, name, parent, text) {
    var known = srv.findFile(srv.normalizeFilename(name));
```

i.e. normalise the filename when finding it on a request.

**Comment by puremourning** [comment]
Actually, if I just normalise the file name always in the keys of `srv.fileMap` it works well, and seems more of a sound approach.

**Comment by marijnh** [comment]
Does attached patch solve your issue? It makes code that has a 'trusted' filename access `fileMap` directly, and makes `findFile` normalize its input.

**Comment by puremourning** [comment]
Yeah it does, though as I mentioned on the PR I'm not 100% confident that it is the right approach or that it is 'safe' in general.

**Comment by marijnh** [comment]
I am. Pretty certain at least ;)

**Comment by puremourning** [comment]
Thanks!!
